python_modules/dagster/dagster/_core/definitions/sensor_definition.py

```diff
@@ -181,6 +181,136 @@ def _get_partition_key_from_event_log_record(event_log_record: "EventLogRecord")
     return None
 
 
+def _get_asset_key_from_event_log_record(event_log_record: "EventLogRecord") -> AssetKey:
+    """
+    Given an event log record, returns the asset key for the event log record if it exists.
+
+    If the asset key does not exist, raises an error.
+    """
+    from dagster._core.storage.event_log.base import EventLogRecord
+
+    check.inst_param(event_log_record, "event_log_record", EventLogRecord)
+
+    dagster_event = event_log_record.event_log_entry.dagster_event
+
+    if dagster_event and dagster_event.asset_key:
+        return dagster_event.asset_key
+
+    raise DagsterInvariantViolationError(
+        "Asset key must exist in event log record passed into _get_asset_key_from_event_log_record"
+    )
+
+
+MAX_NUM_SKIPPED_EVENTS = 25
+
+
+class MultiAssetSensorContextCursor:
+    # Tracks the state of the cursor within the tick, created for utility purposes.
+    # Must call MultiAssetSensorEvaluationContext._update_cursor_after_evaluation at end of tick
+    # to serialize the cursor.
+    def __init__(self, cursor: Optional[str], context: "MultiAssetSensorEvaluationContext"):
+        self._unpacked_cursor: Dict[str, Tuple[Optional[str], Optional[int], List[int]]] = (
+            json.loads(cursor) if cursor else {}
+        )
+        check.dict_param(self._unpacked_cursor, "unpacked_cursor", key_type=str)
+        self._context = context
+        self._cursor_has_been_updated = False
+
+    def get_cursor_for_asset(self, asset_key: AssetKey):
+        partition_key, cursor, skipped_event_ids = self._unpacked_cursor.get(
+            str(asset_key), (None, None, [])
+        )
+        return partition_key, cursor, skipped_event_ids
```

Comment:
Calling this `skipped_event_ids` feels a little off to me, for a few reasons:
* "skip" already has a meaning in the world of sensors, with `SkipReason`, which is independent of the meaning here.
* Partitions are heavily involved in the meaning of what's stored here. It feels like the name should have "partition" in it somewhere.

I'm trying to think through what the best term is for this and what the best way is to think about this.  A useful building block could be a term that means "an event that the user has advanced past".

Here's my understanding of what "skipped events" means: within a particular asset, what's currently called "skipped events" includes all events that occurred before the latest event that the user has advanced past, but that, for one of the asset's partitions, occurred after the latest event that the user has advanced past for that asset partition.

A bit of a mouthful, but maybe that gives us some fodder for a name?

Reply:
Another way of putting it is that we're tracking the latest event that we've advanced passed for every asset partition, except that we're able to safely ignore asset partitions that don't have any events that we haven't advanced past that are before the latest event that we've advanced past for the asset.

(Also a mouthful, just spewing words as I deepen my understanding of what's going on here)

Reply:
I was thinking `unevaluated_event_ids`?

Reply:
> Here's my understanding of what "skipped events" means: within a particular asset, what's currently called "skipped events" includes all events that occurred before the latest event that the user has advanced past, but that, for one of the asset's partitions, occurred after the latest event that the user has advanced past for that asset partition.

Yes, that's right.

If we wanted to include "partition" in the name, we could say, "unevaluated_partition_event_ids"? Though it's possible unevaluated events could be unpartitioned. Can't name a good reason why someone would want to do that, but its still theoretically possible

Reply:
> Though it's possible unevaluated events could be unpartitioned.

IMO we shouldn't support this - there's not a clear criteria for when these should be included.

Reply:
I.e. right now users aren't directly determining which events are included in this data structure, right? We're managing it.